pkg/analysis/passes/manifest/manifest.go

```diff
@@ -1,41 +1,134 @@
 	if err != nil {
-		if os.IsNotExist(err) {
-			pass.ReportResult(pass.AnalyzerName, unsignedPlugin, "unsigned plugin", "MANIFEST.txt file not found. Please refer to the documentation for how to sign a plugin. https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/")
-			return nil, nil
-		} else {
-			if unsignedPlugin.ReportAll {
-				unsignedPlugin.Severity = analysis.OK
-				pass.ReportResult(pass.AnalyzerName, unsignedPlugin, "MANIFEST.txt: plugin is signed", "")
-			}
+		pass.ReportResult(pass.AnalyzerName, unsignedPlugin, "unsigned plugin", "MANIFEST.txt file not found. Please refer to the documentation for how to sign a plugin. https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/")
+		return nil, nil
+	}
+
+	if len(b) == 0 {
+		pass.ReportResult(pass.AnalyzerName, emptyManifest, "empty manifest", "MANIFEST.txt file is empty. Please refer to the documentation for how to sign a plugin. https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/")
+		return nil, nil
+	}
+
+	manifest, err := parseManifestFile(b)
+	if err != nil {
+		pass.ReportResult(pass.AnalyzerName, wrongManifest, "could not parse MANIFEST.txt", "MANIFEST.txt file is not a valid manifest. Please refer to the documentation for how to sign a plugin. https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/")
+		return nil, nil
+	}
+
+	if (manifest.Files == nil) || (len(manifest.Files) == 0) {
+		pass.ReportResult(pass.AnalyzerName, undeclaredFiles, "no files declared in MANIFEST.txt", "No files declared in MANIFEST.txt")
+		return nil, nil
+	}
+
+	// check if all existing files are declared in the manifest
+	_ = filepath.Walk(archiveDir, func(path string, file os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if file.IsDir() {
+			return nil
+		}
+		if file.Name() == "MANIFEST.txt" {
+			return nil
+		}
+		// remove archiveDir from path
+		relativePath := path[len(archiveDir)+1:]
+
+		// check the file is declared
+		if _, ok := manifest.Files[relativePath]; !ok {
+			pass.ReportResult(pass.AnalyzerName, undeclaredFiles, "undeclared files in MANIFEST", fmt.Sprintf("File %s is not declared in MANIFEST.txt", relativePath))
+			return nil
+		}
+
+		// check if the sha256sum is correct
+		sha256sum, err := calculateSha256sum(path)
+		if err != nil {
+			pass.ReportResult(pass.AnalyzerName, undeclaredFiles, "could not calculate sha256sum", fmt.Sprintf("Could not calculate sha256sum for file %s", relativePath))
+			return nil
 		}
-		return nil, err
+
+		if sha256sum != manifest.Files[relativePath] {
+			pass.ReportResult(pass.AnalyzerName, invalidShaSum, "invalid file checksum", fmt.Sprintf("checksum for file %s is invalid", relativePath))
+			return nil
+		}
+		return nil
+	})
+
+	// check if all declared files exist
+	for path := range manifest.Files {
+		if _, err := os.Stat(filepath.Join(archiveDir, path)); os.IsNotExist(err) {
```

Comment:
Thanks for the suggestion and the tip! I like it

I'll change the code to make it neat but I do think this being a CLI tool and the archives containing so little files this might be an over optimization.

Reply:
Agree! Not a huge concern about performance due to the nature of the files